File: java/src/main/java/com/dgarson/claude/sidecar/EventMux.java
package com.dgarson.claude.sidecar;

import claude_sidecar.v1.Sidecar.ServerEvent;
import claude_sidecar.v1.Sidecar.TurnBoundary;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Thread-safe event multiplexer that routes {@link ServerEvent}s to global and per-request
 * subscribers. Mirrors Go's event_mux.go.
 *
 * <p>Global subscribers receive all events. Per-request subscribers receive only events matching
 * their request ID. Per-request subscriptions are automatically closed on TURN_END.</p>
 */
final class EventMux {

    private final EventQueue inbound = new EventQueue();
    private final Object lock = new Object();
    private final Map<String, List<Subscription>> byRequest = new HashMap<>();
    private final List<Subscription> global = new ArrayList<>();
    private boolean closed;
    private final Thread dispatchThread;

    EventMux() {
        dispatchThread = Thread.ofVirtual().name("event-mux-dispatch").start(this::run);
    }

    /** Enqueue an event for dispatch to subscribers. */
    void enqueue(ServerEvent event) {
        inbound.push(event);
    }

    /** Stop accepting events and close all subscriptions. */
    void close() {
        inbound.close();
    }

    /** Subscribe to all events. The returned subscription's queue will receive every event. */
    Subscription subscribeAll(int bufferSize) {
        Subscription sub = new Subscription(bufferSize);
        synchronized (lock) {
            if (closed) {
                sub.close();
                return sub;
            }
            global.add(sub);
        }
        return sub;
    }

    /** Subscribe to events for a specific request ID. */
    Subscription subscribeRequest(String requestId, int bufferSize) {
        Subscription sub = new Subscription(bufferSize);
        synchronized (lock) {
            if (closed) {
                sub.close();
                return sub;
            }
            byRequest.computeIfAbsent(requestId, k -> new ArrayList<>()).add(sub);
        }
        return sub;
    }

    /** Unsubscribe a per-request subscription and close it. */
    void unsubscribeRequest(String requestId, Subscription sub) {
        if (sub == null || requestId == null || requestId.isEmpty()) {
            return;
        }
        synchronized (lock) {
            List<Subscription> subs = byRequest.get(requestId);
            if (subs != null) {
                subs.remove(sub);
                if (subs.isEmpty()) {
                    byRequest.remove(requestId);
                }
            }
        }
        sub.close();
    }

    /** Main dispatch loop, runs on a dedicated thread. */
    private void run() {
        while (true) {
            ServerEvent event = inbound.pop();
            if (event == null) {
                closeAll();
                return;
            }
            List<Subscription> targets = subscriptionsFor(event);
            for (Subscription sub : targets) {
                sub.enqueue(event);
            }
            if (isTurnEnd(event)) {
                closeRequest(event.getRequestId());
            }
        }
    }

    private List<Subscription> subscriptionsFor(ServerEvent event) {
        synchronized (lock) {
            List<Subscription> subs = new ArrayList<>(global);
            String requestId = event.getRequestId();
            if (requestId != null && !requestId.isEmpty()) {
                List<Subscription> requestSubs = byRequest.get(requestId);
                if (requestSubs != null) {
                    subs.addAll(requestSubs);
                }
            }
            return subs;
        }
    }

    private void closeRequest(String requestId) {
        if (requestId == null || requestId.isEmpty()) {
            return;
        }
        List<Subscription> subs;
        synchronized (lock) {
            subs = byRequest.remove(requestId);
        }
        if (subs != null) {
            for (Subscription sub : subs) {
                sub.close();
            }
        }
    }

    private void closeAll() {
        List<Subscription> allSubs;
        synchronized (lock) {
            if (closed) {
                return;
            }
            closed = true;
            allSubs = new ArrayList<>(global);
            for (List<Subscription> list : byRequest.values()) {
                allSubs.addAll(list);
            }
            global.clear();
            byRequest.clear();
        }
        for (Subscription sub : allSubs) {
            sub.close();
        }
    }

    private static boolean isTurnEnd(ServerEvent event) {
        if (!event.hasTurn()) {
            return false;
        }
        return event.getTurn().getKind() == TurnBoundary.Kind.TURN_END;
    }

    // -------------------------------------------------------------------------
    // Inner: thread-safe blocking queue with close semantics
    // -------------------------------------------------------------------------

    static final class EventQueue {
        private final ReentrantLock queueLock = new ReentrantLock();
        private final Condition notEmpty = queueLock.newCondition();
        private final List<ServerEvent> items = new ArrayList<>();
        private boolean queueClosed;

        /** Push an event. Returns false if the queue is closed. */
        boolean push(ServerEvent event) {
            queueLock.lock();
            try {
                if (queueClosed) {
                    return false;
                }
                items.add(event);
                notEmpty.signal();
                return true;
            } finally {
                queueLock.unlock();
            }
        }

        /** Blocking pop. Returns null when the queue is closed and drained. */
        ServerEvent pop() {
            queueLock.lock();
            try {
                while (items.isEmpty() && !queueClosed) {
                    notEmpty.await();
                }
                if (items.isEmpty()) {
                    return null;
                }
                return items.remove(0);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return null;
            } finally {
                queueLock.unlock();
            }
        }

        void close() {
            queueLock.lock();
            try {
                queueClosed = true;
                notEmpty.signalAll();
            } finally {
                queueLock.unlock();
            }
        }
    }

    // -------------------------------------------------------------------------
    // Inner: a subscription backed by a blocking queue, bridged to a
    // LinkedBlockingQueue for consumer convenience.
    // -------------------------------------------------------------------------

    static final class Subscription {
        private final EventQueue queue = new EventQueue();
        private final BlockingQueue<ServerEvent> out;
        private volatile boolean subClosed;

        Subscription(int bufferSize) {
            this.out = new LinkedBlockingQueue<>(Math.max(bufferSize, 1));
            Thread.ofVirtual().name("subscription-pump").start(this::pump);
        }

        private void pump() {
            while (true) {
                ServerEvent event = queue.pop();
                if (event == null) {
                    return;
                }
                try {
                    out.put(event);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }

        void enqueue(ServerEvent event) {
            queue.push(event);
        }

        /**
         * Returns the output queue. Consumers should poll/take from this queue.
         * The queue will stop producing when the subscription is closed.
         */
        BlockingQueue<ServerEvent> queue() {
            return out;
        }

        void close() {
            if (!subClosed) {
                subClosed = true;
                queue.close();
            }
        }

        boolean isClosed() {
            return subClosed;
        }
    }
}

File: go/client/event_mux.go
package client

import (
	"sync"

	pb "github.com/dgarson/claude-sidecar/gen/claude_sidecar/v1"
)

type eventQueue struct {
	mu     sync.Mutex
	cond   *sync.Cond
	items  []*pb.ServerEvent
	closed bool
}

func newEventQueue() *eventQueue {
	queue := &eventQueue{}
	queue.cond = sync.NewCond(&queue.mu)
	return queue
}

func (q *eventQueue) Push(event *pb.ServerEvent) bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	if q.closed {
		return false
	}
	q.items = append(q.items, event)
	q.cond.Signal()
	return true
}

func (q *eventQueue) Pop() (*pb.ServerEvent, bool) {
	q.mu.Lock()
	defer q.mu.Unlock()
	for len(q.items) == 0 && !q.closed {
		q.cond.Wait()
	}
	if len(q.items) == 0 {
		return nil, false
	}
	event := q.items[0]
	q.items = q.items[1:]
	return event, true
}

func (q *eventQueue) Close() {
	q.mu.Lock()
	q.closed = true
	q.cond.Broadcast()
	q.mu.Unlock()
}

type subscription struct {
	queue *eventQueue
	out   chan *pb.ServerEvent
	once  sync.Once
}

func newSubscription(buffer int) *subscription {
	sub := &subscription{
		queue: newEventQueue(),
		out:   make(chan *pb.ServerEvent, buffer),
	}
	go sub.run()
	return sub
}

func (s *subscription) run() {
	for {
		event, ok := s.queue.Pop()
		if !ok {
			close(s.out)
			return
		}
		s.out <- event
	}
}

func (s *subscription) Enqueue(event *pb.ServerEvent) {
	_ = s.queue.Push(event)
}

func (s *subscription) Chan() <-chan *pb.ServerEvent {
	return s.out
}

func (s *subscription) Close() {
	s.once.Do(func() {
		s.queue.Close()
	})
}

type eventMux struct {
	queue     *eventQueue
	mu        sync.Mutex
	byRequest map[string][]*subscription
	global    []*subscription
	closed    bool
}

func newEventMux() *eventMux {
	mux := &eventMux{
		queue:     newEventQueue(),
		byRequest: make(map[string][]*subscription),
	}
	go mux.run()
	return mux
}

func (m *eventMux) Enqueue(event *pb.ServerEvent) {
	_ = m.queue.Push(event)
}

func (m *eventMux) Close() {
	m.queue.Close()
}

func (m *eventMux) SubscribeAll(buffer int) *subscription {
	sub := newSubscription(buffer)
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.closed {
		sub.Close()
		return sub
	}
	m.global = append(m.global, sub)
	return sub
}

func (m *eventMux) SubscribeRequest(requestID string, buffer int) *subscription {
	sub := newSubscription(buffer)
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.closed {
		sub.Close()
		return sub
	}
	m.byRequest[requestID] = append(m.byRequest[requestID], sub)
	return sub
}

func (m *eventMux) UnsubscribeRequest(requestID string, sub *subscription) {
	if sub == nil || requestID == "" {
		return
	}
	m.mu.Lock()
	subs := m.byRequest[requestID]
	for i, item := range subs {
		if item == sub {
			subs = append(subs[:i], subs[i+1:]...)
			break
		}
	}
	if len(subs) == 0 {
		delete(m.byRequest, requestID)
	} else {
		m.byRequest[requestID] = subs
	}
	m.mu.Unlock()
	sub.Close()
}

func (m *eventMux) run() {
	for {
		event, ok := m.queue.Pop()
		if !ok {
			m.closeAll()
			return
		}
		for _, sub := range m.subscriptionsFor(event) {
			sub.Enqueue(event)
		}
		if isTurnEnd(event) {
			m.closeRequest(event.GetRequestId())
		}
	}
}

func (m *eventMux) subscriptionsFor(event *pb.ServerEvent) []*subscription {
	m.mu.Lock()
	defer m.mu.Unlock()
	subs := make([]*subscription, 0, len(m.global))
	subs = append(subs, m.global...)
	requestID := event.GetRequestId()
	if requestID == "" {
		return subs
	}
	return append(subs, m.byRequest[requestID]...)
}

func (m *eventMux) closeRequest(requestID string) {
	if requestID == "" {
		return
	}
	m.mu.Lock()
	subs := m.byRequest[requestID]
	delete(m.byRequest, requestID)
	m.mu.Unlock()
	for _, sub := range subs {
		sub.Close()
	}
}

func (m *eventMux) closeAll() {
	m.mu.Lock()
	if m.closed {
		m.mu.Unlock()
		return
	}
	m.closed = true
	subs := make([]*subscription, 0, len(m.global))
	subs = append(subs, m.global...)
	for _, list := range m.byRequest {
		subs = append(subs, list...)
	}
	m.global = nil
	m.byRequest = make(map[string][]*subscription)
	m.mu.Unlock()
	for _, sub := range subs {
		sub.Close()
	}
}

func isTurnEnd(event *pb.ServerEvent) bool {
	payload, ok := event.Payload.(*pb.ServerEvent_Turn)
	if !ok || payload.Turn == nil {
		return false
	}
	return payload.Turn.Kind == pb.TurnBoundary_TURN_END
}
